File: leetcode/longest_prefix.py
#!/usr/bin/env python3

import sys
import unittest

# number: 14
# section: array/string
# difficulty: easy
# tags: string, trie, top 150

# constraints
# 1 <= strs.length <= 200
# 0 <= strs[i].length <= 200
# strs[i] consists of only lowercase English letters.

# solution: brute-force
# complexity
# run-time: O(n^2)
# space: O(1)
def longest_prefix1(strs):
    if not strs or not len(strs[0]):
        return ''
    elif len(strs) == 1:
        return strs[0]

    prefix = ''
    curr_prefix = ''

    # go over 1st word
    for i in range(len(strs[0])):
        # go over rest of words
        for j in range(1, len(strs)):
            # mismatch
            if i == len(strs[j]) or strs[0][i] != strs[j][i]:
                return prefix

            curr_prefix = strs[0][i]

        prefix += curr_prefix
        curr_prefix = ''

    return prefix if prefix else ''

# solution: trie
# complexity
# run-time: O(n)
# space: O(n)
class Trie:
    def __init__(self):
        self.count = 0
        self.children = {}
        # TODO: store shortest word to use to walk trie?

    def insert(self, word):
        curr = self
        curr.count += 1

        for c in word:
            if c not in curr.children:
                curr.children[c] = Trie()

            curr = curr.children[c]
            curr.count += 1

    # TODO: refactor
    def common_prefix(self):
        curr = self
        count = self.count
        prefix = ""

        while 1:
            if not curr.children:
                break

            # TODO: pick better?
            c = next(iter(curr.children))
            if curr.count != count:
                return prefix[:-1] if prefix else ""

            prefix += c
            curr = curr.children[c]

        return prefix if curr.count == count else prefix[:-1]

def longest_prefix2(strs):
    trie = Trie()

    # build trie
    for w in strs:
        trie.insert(w)

    return trie.common_prefix()

class TestLongestCommonPrefix(unittest.TestCase):

    def test_empty(self):
        strs = []
        self.assertFalse(longest_prefix1(strs))
        self.assertFalse(longest_prefix2(strs))

    def test_single_char(self):
        strs = ['a']
        expected = strs[0]
        self.assertEqual(longest_prefix1(strs), expected)
        self.assertEqual(longest_prefix2(strs), expected)

    def test_two_chars(self):
        strs = ['aa','aa']
        expected = strs[0]
        self.assertEqual(longest_prefix1(strs), expected)
        self.assertEqual(longest_prefix2(strs), expected)

    def test_three_chars(self):
        strs = ['aba','aba']
        expected = strs[0]
        self.assertEqual(longest_prefix1(strs), expected)
        self.assertEqual(longest_prefix2(strs), expected)

    def test_one_char_match(self):
        strs = ['cir','car']
        expected = 'c'
        self.assertEqual(longest_prefix1(strs), expected)
        self.assertEqual(longest_prefix2(strs), expected)

    def test_two_chars_match(self):
        strs = ['flower','flow','flight']
        expected = 'fl'
        self.assertEqual(longest_prefix1(strs), expected)
        self.assertEqual(longest_prefix2(strs), expected)

    def test_no_match(self):
        strs = ['dog','racecar','car']
        self.assertFalse(longest_prefix1(strs))
        self.assertFalse(longest_prefix2(strs))

    def test_no_match(self):
        strs = ['', 'b']
        self.assertFalse(longest_prefix1(strs))
        self.assertFalse(longest_prefix2(strs))

if __name__ == '__main__':
    sys.exit(unittest.main())

File: leetcode/longest_prefix.cpp
#include <iostream>
#include <map>

// number: 14
// title: Longest Common Prefix
// url: https://leetcode.com/problems/longest-common-prefix/
// section: array/string
// difficulty: easy
// tags: string, trie, top 150

// constraints
// 1 <= strs.length <= 200
// 0 <= strs[i].length <= 200
// strs[i] consists of only lowercase English letters.

// complexity
// run-time: O(n)
// space: O(1)
class Trie {
public:
    // need map, we'll be iterating over keys
    using ChildrenMap = std::map<char, Trie>;

    Trie()
    : d_count(0) {}

    void insert(const std::string& word) {
        Trie* node = this;
        node->count() += 1;

        for (const auto& c: word) {
            ChildrenMap& children = node->d_children;
            auto iter = children.find(c);
            if (iter == children.end()) {
                children[c] = Trie();
            }

            node = &node->d_children[c];
            node->count() += 1;
        }
    }

    std::string commonPrefix() {
        Trie* node = this;
        std::string prefix;
        auto iter = node->d_children.begin();

        while (1) {
            if (node->d_children.empty()) {
                break;
            }

            // TODO: figure out iterator
            auto c = iter->first;
            ++iter;
            //if (iter == node->children().end()) {
            //    std::cout << "no children" << std::endl;
            //    break;
            //}

            if (node->count() != d_count) {
                if (!prefix.empty()) {
                    std::cout << "prefix:" << prefix << std::endl;
                    return prefix.substr(0, prefix.size()-1);
                } else {
                    std::cout << "empty prefix" << std::endl;
                    return "";
                }
            }

            prefix += c;
            node = &node->d_children[c];
        }

        if (node->count() != d_count) {
            std::cout << "prefix:" << prefix << std::endl;
            return prefix.substr(0, prefix.size()-1);
        }

        std::cout << "prefix:" << prefix << std::endl;

        return prefix;
    }

    int& count() { return d_count; }
    void count(int value) { d_count = value; }

private:
    ChildrenMap d_children;
    int         d_count;
};

// WIP solution: trie
// complexity:
// run-time: O(n)
// space: O(n)
std::string longestCommonPrefix(const std::vector<std::string>& strs) {
    Trie trie;

    for (const auto& w: strs) {
        trie.insert(w);
    }

    return trie.commonPrefix();
}
